src/inpod/admin.rs

```diff
@@ -0,0 +1,163 @@
+use hyper::{Request, Response};
+use serde::{Deserialize, Serialize};
+use tracing::error;
+
+use std::collections::HashMap;
+use std::sync::RwLock;
+
+#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
+pub enum State {
+    Pending,
+    Up,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ProxyState {
+    pub state: State,
+
+    // using refernce counts to account for possible race between the proxy task that notifies us
+    // that a proxy is down, and the proxy factory task that notifies us when it is up.
+    #[serde(skip)]
+    count: usize,
+}
+
+#[derive(Default)]
+pub struct WorkloadManagerAdminHandler {
+    state: RwLock<HashMap<String, ProxyState>>,
+}
+
+impl WorkloadManagerAdminHandler {
+    pub fn proxy_pending(&self, uid: &str) {
+        let mut state = self.state.write().unwrap();
+
+        // don't increment count here, as it is only for up and down. see comment in count.
+        match state.get_mut(uid) {
+            Some(key) => {
+                key.state = State::Pending;
+            }
+            None => {
+                state.insert(
+                    uid.to_string(),
+                    ProxyState {
+                        state: State::Pending,
+                        count: 0,
+                    },
+                );
+            }
+        }
+    }
+    pub fn proxy_up(&self, uid: &str) {
+        let mut state = self.state.write().unwrap();
+
+        match state.get_mut(uid) {
+            Some(key) => {
+                key.count += 1;
+                key.state = State::Up;
+            }
+            None => {
+                state.insert(
+                    uid.to_string(),
+                    ProxyState {
+                        state: State::Up,
+                        count: 1,
+                    },
+                );
+            }
+        }
+    }
+
+    pub fn proxy_down(&self, uid: &str) {
+        let mut state = self.state.write().unwrap();
+
+        match state.get_mut(uid) {
+            Some(key) if key.count > 0 => {
+                key.count -= 1;
+                if key.count == 0 {
+                    state.remove(uid);
+                }
+            }
+            _ => {
+                error!("proxy_down called where no proxy was created");
+                debug_assert!(false, "proxy_down called where no proxy was created");
+            }
+        }
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
+        let serialized_state = if let Ok(state) = self.state.read() {
+            serde_json::to_vec(&*state)
+        } else {
+            error!("Failed to read state");
+            return vec![];
+        };
+        match serialized_state {
+            Ok(v) => v,
+            Err(e) => {
+                error!("Failed serializing state {:?}", e);
+                vec![]
+            }
+        }
+    }
+}
+
+impl crate::admin::AdminHandler for WorkloadManagerAdminHandler {
+    fn path(&self) -> &'static str {
+        "/workloadmanager"
```

Comment:
My reasoning for the new interface:
- Its only available if inpod mode is on . It's dynamically added to the admin interface.
- Another benefit is that all the logic related to "inpod" is inside the inpod module (I feel this makes the code easier  to grok).

Reply:
I like the separation (and probably a good idea to use it in other plaecs!), but maybe we can just make it so keys under config_dump are dynamically loaded?

Reply:
I'm not sure how do this dynamically without an interface?
what do you mean by "keys"?

Reply:
sorry for late response - just saw this comment.

I think I'm misunderstanding your intention here - Does each handler returns its key and value? wouldn't this require an interface too for the handlers?

Reply:
It could return a `map<string, serde::RawValue>` or similar?

Or just a RawValue and have a `key()` method (then it can only write to 1 key, though).

(RawValue may not be the right type either, not sure)

Reply:
we can do that, but if we want to use it in other places it might be useful to have it return a future, for async invocations. for example, `handle_pprof` must be async